contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +695,93 @@ with minimal disruption to users would be:
+on probed plugins need to be audited and modified to handle nil plugins returned
+as detailed below.
+
+### Enhancements in ProbeVolumePlugins
+
+A new file pkg/volume/util/csimigration/plugins.go will have the following functions:
+```
+func ProbePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+func ProbeAttachablePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+```
+They will [1] check plugin specific migration feature flags as well as [2] the overall
+CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+of a legacy in-tree plugin will get invoked. Kubelet's ProbeVolumePlugins and Kube
+Controller Manager's ProbeAttachableVolumePlugins, ProbeExpandableVolumePlugins and
+ProbeControllerVolumePlugins will call the above instead of directly probing in-tree plugins
+that will be migrated.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+pkg/volume/util/csimigration/plugins.go so that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on a common utility function:
+```
+func IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+```
+that will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of nil Plugin Object returned by FindPluginBySpec/FindPluginByName
+
+Once an in-tree plugin is no longer probed, all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return a nil plugin
+object to code paths that invoke those functions. For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return nil for a migrated
+plugin that is no longer probed in ProbeControllerVolumePlugins.
+2. FindExpandablePluginBySpec in the expand controller will return nil for a migrated
+plugin that is no longer probed in ProbeExpandableVolumePlugins.
+3. FindAttachablePluginBySpec in the attach/detach controller will return nil for
+a migrated plugin that is no longer probed in ProbeAttachableVolumePlugins.
+
+Each such code path will need to be audited and modified such that:
+1. Code preceding the invocation of Find*Plugin routines performs the necessary
+translation from in-tree to CSI so that when a Find*Plugin routine is invoked,
+it is passed a translated CSI PV rather than an in-tree PV.
+2. Upon receiving a nil plugin, the code checks migration status of a plugin based
```

Comment:
technically it should have been translated/migrated already and we should be trying to get the CSI Plugin instead of the nil in-tree plugin. If we ever try to get the nil in-tree plugin that seems like a real error, no?

Reply:
Exactly. If [1] is done completely and correctly, [2] should not be necessary for majority of the cases. I will clarify this.